ASSERTION FAILED: isPublicSuffixCF(publicSuffix) in PublicSuffixStore::addPublicSuffix

szewai opened the pull request:
#### 4031410eeeb157432f0a6e0ca143e54d02665096
<pre>
ASSERTION FAILED: isPublicSuffixCF(publicSuffix) in PublicSuffixStore::addPublicSuffix
<a href="https://bugs.webkit.org/show_bug.cgi?id=274418">https://bugs.webkit.org/show_bug.cgi?id=274418</a>
<a href="https://rdar.apple.com/128255984">rdar://128255984</a>

Reviewed by Chris Dumez.

The assertion was introduced in 276834@main to help debug <a href="https://rdar.apple.com/125417343">rdar://125417343</a>, where we suspected the cause is unexpected
entry being added to public suffix cache, like 276635@main. However, because UI process and web process have different
sandbox access, isPublicSuffixCF (_CFHostIsDomainTopLevel) might return different values on the same string. That means,
in web process, the assertion would be hit even though the string is a valid public suffix according to UI process&apos;s
check.

As the original goal is to ensure UI process only sends valid public suffix (or empty string) to web process, the patch
does the hardening by introducing a new class PublicSuffix. UI process needs to explicity contruct it from URL with
PublicSuffixStore and send PublicSuffix object to web process, which is less error-prone than allowing UI process sends
public suffix in string. With this change, we no longer need the assertion.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/PublicSuffix.h: Added.
(WebCore::PublicSuffix::fromRawString):
(WebCore::PublicSuffix::isValid const):
(WebCore::PublicSuffix::string const):
(WebCore::PublicSuffix::isolatedCopy const):
(WebCore::PublicSuffix::PublicSuffix):
(WebCore::PublicSuffix::operator== const):
(WebCore::PublicSuffix::isHashTableDeletedValue const):
(WebCore::PublicSuffix::hash const):
(WebCore::PublicSuffix::PublicSuffixHash::hash):
(WebCore::PublicSuffix::PublicSuffixHash::equal):
* Source/WebCore/platform/PublicSuffixStore.cpp:
(WebCore::PublicSuffixStore::publicSuffix const):
* Source/WebCore/platform/PublicSuffixStore.h:
(): Deleted.
* Source/WebCore/platform/cocoa/PublicSuffixStoreCocoa.mm:
(WebCore::PublicSuffixStore::platformIsPublicSuffix const):
(WebCore::PublicSuffixStore::enablePublicSuffixCache):
(WebCore::PublicSuffixStore::addPublicSuffix):
* Source/WebKit/Shared/GoToBackForwardItemParameters.h:
* Source/WebKit/Shared/GoToBackForwardItemParameters.serialization.in:
* Source/WebKit/Shared/LoadParameters.h:
* Source/WebKit/Shared/LoadParameters.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/Tests/WebCore/PublicSuffix.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: <a href="https://commits.webkit.org/279128@main">https://commits.webkit.org/279128@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ba11f571fd1c3b0640a985b070bcb23a982accc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52524 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31858 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4948 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55798 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3247 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54829 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38394 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2947 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42686 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2067 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54621 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29526 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45326 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23775 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26685 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2602 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1406 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48577 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2756 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57394 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27656 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2778 "Found 6 new test failures: imported/w3c/web-platform-tests/css/css-view-transitions/animating-new-content.html, imported/w3c/web-platform-tests/css/css-view-transitions/exit-transition-with-anonymous-layout-object.html, imported/w3c/web-platform-tests/css/css-view-transitions/massive-element-left-of-viewport-offscreen-new.html, imported/w3c/web-platform-tests/css/css-view-transitions/span-with-overflowing-text-and-box-decorations.html, imported/w3c/web-platform-tests/css/css-view-transitions/span-with-overflowing-text-hidden.html, imported/w3c/web-platform-tests/css/css-view-transitions/transform-origin-view-transition-group.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50083 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28885 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45444 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49348 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11487 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29796 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28633 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2767,6 +2767,14 @@ TextStream& operator<<(TextStream& ts, const WebProcessProxy& process)
+{
+    bool isPublicSuffix = PublicSuffixStore::singleton().isPublicSuffix(publicSuffix);
+    if (!isPublicSuffix)
+        WEBPROCESSPROXY_RELEASE_LOG_ERROR(Network, "handleUnrecognizedPublicSuffix: Received an recognized public suffix %" PUBLIC_LOG_STRING, publicSuffix.utf8().data());
```
Isn't this going to log domains on customer builds?

szewai replied:
This should be TLD (`com`, `net`, etc) not eTLD + 1 (`apple.com`), so I don't think it is privacy sensitive. 
I can change it to private if you think it should be.

cdumez replied:
What is the purpose of this patch? Why are we trying to validate the public suffix at all?

The code does this:
1. UIProcess sends public suffix to the WebProcess
2. Before adding to PublicSuffixStore, WebProcess checks if public suffix is valid
3. If it is not valid, it asks the UIProcess to validate it

Seems silly. Why ask the UIProcess to validate something it sent us?

szewai replied:
The assertion was introduced in [276834@main](https://commits.webkit.org/276834@main) to help debug rdar://125417343, where we suspected the cause is web process adding unexpected item to public suffix cache. 
There was a bug where UI process sends topPrivatelyControlledDomain instead of publicSuffix [276635@main](https://commits.webkit.org/276635@main).

This patch aims to replace the assertion in web process. If we believe everything works correctly, then indeed we don't need the validation process.

szewai replied:
Thinking about it, maybe we could wrap the public suffix in a class and send that instead of a plain string in message instead. Do you think it's better to do this?

cdumez replied:
Yes, having a PublicSuffix class makes sense to me.

cdumez replied:
If you want I can do that too, if you want to go back to site isolation.

szewai replied:
Seems not too complained; I will work on it

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2767,6 +2767,14 @@ TextStream& operator<<(TextStream& ts, const WebProcessProxy& process)
+    bool isPublicSuffix = PublicSuffixStore::singleton().isPublicSuffix(publicSuffix);
+    if (!isPublicSuffix)
+        WEBPROCESSPROXY_RELEASE_LOG_ERROR(Network, "handleUnrecognizedPublicSuffix: Received an recognized public suffix %" PUBLIC_LOG_STRING, publicSuffix.utf8().data());
+    MESSAGE_CHECK(isPublicSuffix);
```
What is the purpose of killing the child process here? A compromised WebProcess could easily NOT send this IPC..

szewai replied:
This is for collecting reports about unrecognized public suffix -- same as what the current assertion does.

cdumez replied:
MESSAGE_CHECK() doesn't generate a crash log though.

cdumez replied:
> same as what the current assertion does

Yes, I am questioning this assertion too. The UIProcess sends the WebProcess a Public Suffix, why is the WebProcess asserting that what it received is a public suffix? What's the purpose? Why can't the WebProcess trust the UIProcess? Also, clearly, the WebProcess is not even able to do the validation anyway.

cdumez commented on Source/WebCore/platform/PublicSuffixStore.h:
```diff
@@ -38,14 +38,28 @@ class PublicSuffixStore {
     WEBCORE_EXPORT void clearHostTopPrivatelyControlledDomainCache();
+    WEBCORE_EXPORT void enablePublicSuffixCache();
 
+    class PublicSuffix {
```
Maybe we can define this *outside* of the PublicSuffixStore scope? I don't feel strongly but it feels like this might be a bit nicer.

cdumez replied:
Odds are we may want to add it to its own header too, with its own HashTraits and isolatedCopy().

cdumez commented on Source/WebCore/platform/PublicSuffixStore.h:
```diff
@@ -38,14 +38,28 @@ class PublicSuffixStore {
+        String string() const { return m_string; }
+
+    private:
+        explicit PublicSuffix(const String& string) : m_string(string) { }
```
fromRawString() passes a r-value reference but this doesn't and it doesn't WTFMove(), why not?

cdumez commented on Source/WebCore/platform/PublicSuffixStore.h:
```diff
@@ -38,14 +38,28 @@ class PublicSuffixStore {
+        }
+        PublicSuffix() = default;
+        bool isValid() const { return !m_string.isEmpty(); }
+        String string() const { return m_string; }
```
should return a `const String&`.

cdumez commented on Source/WebCore/platform/cocoa/PublicSuffixStoreCocoa.mm:
```diff
@@ -60,28 +60,25 @@ static bool isPublicSuffixCF(const String& domain)
     Locker locker { m_publicSuffixCacheLock };
     ASSERT(m_publicSuffixCache);
-    m_publicSuffixCache->add(crossThreadCopy(publicSuffix));
+    m_publicSuffixCache->add(crossThreadCopy(publicSuffix.string()));
```
This isn't great. The HashMap should contain PublicSuffix objects ideally.

cdumez commented on Source/WebCore/platform/cocoa/PublicSuffixStoreCocoa.mm:
```diff
@@ -42,7 +42,8 @@ static bool isPublicSuffixCF(const String& domain)
     {
         Locker locker { m_publicSuffixCacheLock };
-        if (m_publicSuffixCache && m_publicSuffixCache->contains(domainString))
+        auto publicSuffix = PublicSuffix::fromRawString(String { domainString });
```
Seems unfortunate to do this when `!m_publicSuffixCache`

szewai committed:
ASSERTION FAILED: isPublicSuffixCF(publicSuffix) in PublicSuffixStore::addPublicSuffix
https://bugs.webkit.org/show_bug.cgi?id=274418
rdar://128255984

Reviewed by Chris Dumez.

The assertion was introduced in 276834@main to help debug rdar://125417343, where we suspected the cause is unexpected
entry being added to public suffix cache, like 276635@main. However, because UI process and web process have different
sandbox access, isPublicSuffixCF (_CFHostIsDomainTopLevel) might return different values on the same string. That means,
in web process, the assertion would be hit even though the string is a valid public suffix according to UI process's
check.

As the original goal is to ensure UI process only sends valid public suffix (or empty string) to web process, the patch
does the hardening by introducing a new class PublicSuffix. UI process needs to explicity contruct it from URL with
PublicSuffixStore and send PublicSuffix object to web process, which is less error-prone than allowing UI process sends
public suffix in string. With this change, we no longer need the assertion.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/PublicSuffix.h: Added.
(WebCore::PublicSuffix::fromRawString):
(WebCore::PublicSuffix::isValid const):
(WebCore::PublicSuffix::string const):
(WebCore::PublicSuffix::isolatedCopy const):
(WebCore::PublicSuffix::PublicSuffix):
(WebCore::PublicSuffix::operator== const):
(WebCore::PublicSuffix::isHashTableDeletedValue const):
(WebCore::PublicSuffix::hash const):
(WebCore::PublicSuffix::PublicSuffixHash::hash):
(WebCore::PublicSuffix::PublicSuffixHash::equal):
* Source/WebCore/platform/PublicSuffixStore.cpp:
(WebCore::PublicSuffixStore::publicSuffix const):
* Source/WebCore/platform/PublicSuffixStore.h:
(): Deleted.
* Source/WebCore/platform/cocoa/PublicSuffixStoreCocoa.mm:
(WebCore::PublicSuffixStore::platformIsPublicSuffix const):
(WebCore::PublicSuffixStore::enablePublicSuffixCache):
(WebCore::PublicSuffixStore::addPublicSuffix):
* Source/WebKit/Shared/GoToBackForwardItemParameters.h:
* Source/WebKit/Shared/GoToBackForwardItemParameters.serialization.in:
* Source/WebKit/Shared/LoadParameters.h:
* Source/WebKit/Shared/LoadParameters.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/Tests/WebCore/PublicSuffix.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: https://commits.webkit.org/279128@main
